File: php/src/Support/ActivityText.php
<?php

declare(strict_types=1);

namespace App\Support;

/**
 * Renders one activity row as a plain, readable sentence for the feed. A line-for-line PHP port
 * of the reference app's `ACTIVITY_LABELS` (src/types/activity.ts) plus the actor-name prefix the
 * reference's `ActivityFeed.tsx` renders around it.
 */
final class ActivityText
{
    private const LABELS = [
        'booking_confirmed' => 'booked (confirmed)',
        'booking_waitlisted' => 'joined the waitlist',
        'booking_cancelled' => 'cancelled their booking',
        'booking_promoted' => 'was promoted from the waitlist',
        'checked_in' => 'checked in',
        'event_created' => 'created this event',
        'event_updated' => 'updated this event',
    ];

    /** @param array<string, mixed> $entry */
    public static function describe(array $entry): string
    {
        $who = is_string($entry['actorName'] ?? null) && $entry['actorName'] !== '' ? $entry['actorName'] : 'Someone';
        $action = is_string($entry['action'] ?? null) ? $entry['action'] : '';
        $label = self::LABELS[$action] ?? 'did something';

        return "{$who} {$label}";
    }
}

File: mobile-flutter/lib/core/activity_text.dart
import '../models/activity_entry.dart';

/// Renders one activity row as a plain, readable sentence for the feed.
/// Mirrors `ACTIVITY_LABELS` in `web/src/types/activity.ts` (the web app
/// renders `"{actorName} {label}"` directly in `ActivityFeed.tsx`), with one
/// addition: an `ActivityAction.unknown` arm for forward compatibility with
/// a server-sent `action` value this client doesn't recognize yet.
String describeActivity(ActivityEntry entry) {
  final who = entry.actorName.isNotEmpty ? entry.actorName : 'Someone';
  final label = switch (entry.action) {
    ActivityAction.bookingConfirmed => 'booked (confirmed)',
    ActivityAction.bookingWaitlisted => 'joined the waitlist',
    ActivityAction.bookingCancelled => 'cancelled their booking',
    ActivityAction.bookingPromoted => 'was promoted from the waitlist',
    ActivityAction.checkedIn => 'checked in',
    ActivityAction.eventCreated => 'created this event',
    ActivityAction.eventUpdated => 'updated this event',
    ActivityAction.unknown => 'did something on this event',
  };
  return '$who $label';
}
